projects/lxc/fuzz-lxc-config-read.c

```diff
@@ -0,0 +1,41 @@
+/*
+# Copyright 2021 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+*/
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "conf.h"
+#include "confile.h"
+#include "utils.h"
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+	int fd = -1;
+	char tmpf[] = "fuzz-lxc-config-read-XXXXXX";
+	struct lxc_conf *conf = NULL;
+
+	fd = lxc_make_tmpfile(tmpf, false);
+	lxc_write_nointr(fd, data, size);
+	close(fd);
+
+	conf = lxc_conf_init();
+	lxc_config_read(tmpf, conf, false);
+	lxc_conf_free(conf);
```

Comment:
@brauner this line triggers ASan from time to time with something like:
```
==13==ERROR: AddressSanitizer: heap-use-after-free on address 0x6120000266f0 at pc 0x0000005c7bfd bp 0x7ffe9a3f1570 sp 0x7ffe9a3f1568
READ of size 8 at 0x6120000266f0 thread T0
SCARINESS: 51 (8-byte-read-heap-use-after-free)
    #0 0x5c7bfc in lxc_free_netdev /src/lxc/src/lxc/confile_utils.c:414:15
    #1 0x5c7300 in lxc_free_networks /src/lxc/src/lxc/confile_utils.c:482:3
    #2 0x59cd2a in lxc_conf_free /src/lxc/src/lxc/conf.c:3961:2
    #3 0x581c0b in LLVMFuzzerTestOneInput /src/fuzz-lxc-config-read.c:19:2
    #4 0x484783 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:599:15
    #5 0x483c8a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:505:3
    #6 0x4859e7 in fuzzer::Fuzzer::MutateAndTestOne() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:745:19
    #7 0x486575 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:883:5
    #8 0x474307 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:906:6
    #9 0x4a0162 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #10 0x7f8900f9f83f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)
    #11 0x449278 in _start (/out/fuzz-lxc-config-read+0x449278)
```
I'm not sure whether it's an lxc bug or a bug in the fuzz target. I kind of assumed that it's always safe to call `lxc_conf_free` right after `lxc_config_read` regardless of whether `lxc_config_read` has failed or not.

Reply:
This looks like a legitimate bug!
I sent a PR to fix this!